compiler/parser/RdrHsSyn.hs

```diff
@@ -2852,14 +2866,74 @@ mkTemplateDecls header fields decls = do
   choiceDataDecls <- concat <$> traverse mkChoiceDataDecls vtChoices
   let templateName = occNameString . rdrNameOcc <$> vtTemplateName
       templateDataDecl = mkTemplateDataDecl (combineLocs vtTemplateName fields) vtTemplateName vtTypeVars ci
+      (letDecl,sharedBinds) = shareTemplateLetBindings conName vtLetBindings
       choicesWithArchive = mkArchiveChoice : vtChoices
-      templateInstances = mkTemplateInstanceDecl templateName conName vt
+      templateInstances = mkTemplateInstanceDecl sharedBinds templateName conName vt
       choiceInstanceDecls = concatMap (mkChoiceInstanceDecl templateName vtTypeVars) choicesWithArchive
-      choiceDecls = concatMap (mkChoiceDecls (getLoc templateName) conName vtLetBindings) choicesWithArchive
-      keyInstanceDecl = mkKeyInstanceDecl templateName conName vt -- <$> kdKeyType . unLoc <$> vtKeyData
-  return $ toOL $ templateDataDecl : choiceDataDecls
+      choiceDecls = concatMap (mkChoiceDecls (getLoc templateName) conName sharedBinds) choicesWithArchive
+      keyInstanceDecl = mkKeyInstanceDecl sharedBinds templateName conName vt -- <$> kdKeyType . unLoc <$> vtKeyData
+
+  return $ toOL $ templateDataDecl : choiceDataDecls ++ [letDecl]
                ++ templateInstances ++ (choiceInstanceDecls ++ choiceDecls ++ keyInstanceDecl)
 
+
+shareTemplateLetBindings :: Located RdrName -> LHsLocalBinds GhcPs -> (LHsDecl GhcPs, LHsLocalBinds GhcPs)
+shareTemplateLetBindings conName vtLetBindings = (letDecl,sharedBinds)
+  where
+    letFnName :: RdrName
+    letFnName = mkRdrUnqual $ mkVarOcc ("_templateLet_" ++ rdrNameToString conName)
+
+    vars :: [RdrName]
+    vars = varsOfBinds vtLetBindings
+
+    letDecl :: LHsDecl GhcPs
+    letDecl = functionBind letFnName defArgs (noLoc $ HsLet noExt binds body)
+      where
+        defArgs = [this]
+        this = asPatRecWild "this" conName
+        binds = extendLetBindings vtLetBindings (dummyBinds defArgs)
+
+        body :: LHsExpr GhcPs
+        body = mkTupleExp (map mkRdrExp vars)
+
+    sharedBinds :: LHsLocalBinds GhcPs
+    sharedBinds = extendLetBindings (mkLetBindings (listToBag [bind])) (dummyBinds (map mkRdrPat vars))
+      where
+        bind :: LHsBind GhcPs
+        bind = noLoc $ PatBind noExt lhs rhs ([], [])
+
+        lhs :: Pat GhcPs
+        lhs = mkTuplePat (map mkRdrPat vars)
+
+        rhs :: GRHSs GhcPs (LHsExpr GhcPs)
+        rhs = GRHSs noExt [noLoc $ GRHS noExt [] exp] (noLoc emptyLocalBinds)
+
+        exp :: LHsExpr GhcPs
+        exp = mkApp (mkRdrExp letFnName) (mkUnqualVar $ mkVarOcc "this")
+
+
+functionBind :: RdrName -> [Pat GhcPs] -> LHsExpr GhcPs -> LHsDecl GhcPs
```

Comment:
This looks reasonably close to `makeFunBind` or `funBind`. Can we unify the 3 somehow instead of adding another one?

Reply:
Hmm.
It's different from `makeFunBind`
and `funBind` is just a very shallow wrapper around `FunBind`

However, I made a small change to `funBind` to allow it to be reused by `functionBind`